packages/sources/coinmarketcap-test/src/crypto-utils.ts

```diff
@@ -0,0 +1,261 @@
+    required: false,
+    type: 'string',
+  },
+} as const
+
+interface PriceInfo {
+  price: number
+  volume_24h: number
+  percent_change_1h: number
+  percent_change_24h: number
+  percent_change_7d: number
+  percent_change_30d: number
+  market_cap: number
+}
+
+export interface ProviderResponseBody {
+  data: {
+    [key: string]: {
+      id: number
+      name: string
+      symbol: string
+      slug: string
+      is_active: number
+      is_fiat: number
+      circulating_supply: number
+      total_supply: number
+      max_supply: number
+      date_added: string
+      num_market_pairs: number
+      cmc_rank: number
+      last_updated: string
+      tags: string[]
+      platform: string
+      quote: {
+        [key: string]: PriceInfo
+      }
+    }
+  }
+  status: {
+    timestamp: string
+    error_code: number
+    error_message: string
+    elapsed: number
+    credit_count: number
+  }
+  cost: number
+}
+
+export type CryptoEndpointTypes = {
+  Request: {
+    Params: CryptoRequestParams
+  }
+  Response: SingleNumberResultResponse
+  CustomSettings: typeof customSettings
+  Provider: {
+    RequestBody: never
+    ResponseBody: ProviderResponseBody
+  }
+}
+
+// Coin IDs fetched from the ID map: https://coinmarketcap.com/api/documentation/v1/#operation/getV1CryptocurrencyMap
+const presetIds: { [symbol: string]: number } = {
+  '1INCH': 8104,
+  AAVE: 7278,
+  BAL: 5728,
+  BAT: 1697,
+  BCH: 1831,
+  BNB: 1839,
+  BNT: 1727,
+  BTC: 1,
+  COMP: 5692,
+  CRO: 3635,
+  CRV: 6538,
+  ETC: 1321,
+  ETH: 1027,
+  FNX: 5712,
+  FTT: 4195,
+  HT: 2502,
+  KCS: 2087,
+  KLAY: 4256,
+  KNC: 1982,
+  LEO: 3957,
+  LINK: 1975,
+  MIM: 162,
+  MKR: 1518,
+  OHM: 16209,
+  OHMV2: 9067,
+  OKB: 3897,
+  RCN: 2096,
+  REN: 2539,
+  SNX: 2586,
+  STETH: 8085,
+  SUSHI: 6758,
+  UMA: 5617,
+  UNI: 7083,
+  WOM: 5328,
+  YFI: 5864,
+}
+
+type Payload = { payload: CryptoRequestParams; id?: string }
+
+const chunkArray = (params: string[], size = 120): string[][] =>
+  params.length > size ? [params.slice(0, size), ...chunkArray(params.slice(size), size)] : [params]
+
+const groupByBaseOptions = (params: CryptoRequestParams[]) => {
+  const id: Payload[] = []
+  const slug: Payload[] = []
+  const symbol: Payload[] = []
+  const group: { id?: Payload[]; slug?: Payload[]; symbol?: Payload[] } = {}
+  params.forEach((param) => {
+    let optionContainer = []
+    // We keep original params in payload so that we  don't modify them when there is base-to-id override
+    const input: { payload: CryptoRequestParams; id?: string } = { payload: param }
+
+    if (input.payload.base && presetIds[input.payload.base]) {
+      input.id = presetIds[input.payload.base].toString()
+    }
+
+    if (input.id || input.payload.cid) {
+      optionContainer = id
+    } else if (input.payload.slug) {
+      optionContainer = slug
+    } else if (input.payload.base) {
+      optionContainer = symbol
+    }
+    optionContainer.push(input)
```

Comment:
Could we just `.push()` into the respective array directly here? This seems a bit over-complicated